### PR TITLE
ESP8266Server class

### PR DESCRIPTION
**Pull request by lasselukkari**
Currently the usage of ESP8622Client is hard  when runnig as server because the ESP8622Client writes to the default channel if the channel id is not specified in the constructor. 

The new ESP8622Server class returns a ESP8622Client instance with the correct channel id set when data is available.

You probably had something similar already in mind. Please let me know if you want me to change something.

### CODE REVIEW

File: ESP8266Server.cpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#include "ESP8266Server.h"
+
+void ESP8266Server::begin()
+{
+    for (unsigned int i = 0; i < 5; i++){
```
Comment (Diaoul) [inline ESP8266Server.cpp]: Use [ESP8266_MAX_CONNECTIONS](https://github.com/Diaoul/arduino-ESP8266/blob/master/ESP8266.h#L9) here

File: ESP8266Server.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+class ESP8266Server
+{
+public:
+    ESP8266Server(ESP8266& esp8266, uint16_t port) : _esp8266(esp8266), _port(port) {}
+    ESP8266Client * available();
+    virtual void begin();
+
+protected:
+    uint16_t _port;
+    ESP8266Client * _clients[5];
```
Comment (Diaoul) [inline ESP8266Server.h]: Same as above

**Comment by lasselukkari** [comment]
Yes, it creates 5 instances. That is the whole idea. You may have 5 different incoming connectios that you are handling at the same time. I can't see how you could implement the fuctionality in any other way if you want to keep it compatible with the Client interface.

**Comment by tprochazka** [comment]
You can create new instance on demand, in most cases will not be 5 incoming connections.

**Comment by lasselukkari** [comment]
The Client class has one pointer to the esp8266 instance, one boolean value and one int. I don't see the overhead as a problem. Generally dynamic memory allocation is a bad idea with Arduino although i don't think it would really be problem here.

In my projects when is use the ESP with a webserver usually the first page load creates atleast 3 simultaneus requests: html, favico and a rest api request.

**Comment by jonathanchristison** [comment]
@lasselukkari I think I fixed the issues you where having by adding a bunch of overloads for char arrays and string types, its a bit unclean but it seems to do the job.

The fault was a oversight on my behalf, sizeof() does not give the size of a string unless its a string literal. I had only tested this with a string literal. 

See pull request 13 - https://github.com/Diaoul/arduino-ESP8266/pull/13

**Comment by Diaoul** [comment]
@lasselukkari: please add an example in `examples/`

**Comment by lasselukkari** [comment]
Is it ok if the example requires one additional library?

**Comment by Diaoul** [comment]
It is OK as long as it is a known library and covers a common use case I guess.

**Comment by tprochazka** [comment]
> In my projects when is use the ESP with a webserver usually the first page load creates at least 3 simultaneous requests: html, favico and a rest api request.

So. I still don't understand one thing. How you use it. You have 5 instances which are able send data to different channel, but all of them read the same data, data from all incoming requests. So if you have 3 simultaneous incoming connection read() will read all of them as one stream? How do you handle it? Yes you can search for something like "GET " and url after it.... This is the same situation as we discuss in my pull request with multiple Stream implementation and ESP8622Client  is Stream implemention...

**Comment by lasselukkari** [comment]
I don't think it's possible handle 5 different simultaneus connections unless you read them one by one. The idea of the Server class is to return a connected Client instance that has the correct id set so that you can write back to the correct channel by using the same client. 

It doesn't try to solve any of the problems with the ESP8266 multiplexing.

**Comment by tprochazka** [comment]
It is not necessary to read 5 connection simultaneous, ESP8266 serialize them, it comes one by one, so I think that we actually can have 5 different streams, only one will be available at one time. But also solution with one stream which will be possible to read by one IPD block is enough I think.

**Pull request closed without merging**